src/NuGet.ApplicationInsights.Owin/ApplicationInsightsMiddleware.cs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.Owin;
+
+namespace NuGet.ApplicationInsights.Owin
+{
+    public class ApplicationInsightsMiddleware
+        : OwinMiddleware
+    {
+        public const string OwinRequestIdKey = "owin.RequestId";
+
+        private readonly TelemetryClient _telemetryClient;
+
+        public ApplicationInsightsMiddleware(OwinMiddleware next)
+            : this(next, null)
+        {
+        }
+
+        public ApplicationInsightsMiddleware(OwinMiddleware next, TelemetryConfiguration telemetryConfiguration) 
+            : base(next)
+        {
+            _telemetryClient = telemetryConfiguration == null 
+                ? new TelemetryClient()
+                : new TelemetryClient(telemetryConfiguration);
+        }
+
+        public override async Task Invoke(IOwinContext context)
+        {
+            var requestId = context.Get<string>(OwinRequestIdKey);
+
+            var requestMethod = context.Request.Method;
+            var requestPath = context.Request.Path.ToString();
+            var requestUri = context.Request.Uri;
+
+            var requestStartDate = DateTimeOffset.Now;
+            var stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            var requestFailed = false;
+
+            try
+            {
+                OwinRequestIdContext.Set(requestId);
+
+                if (Next != null)
+                {
+                    await Next.Invoke(context);
+                }
+            }
+            catch (Exception)
+            {
+                requestFailed = true;
```

Comment:
That will help. However, if we have a different middleware that explicitly handles exceptions, it might be a concern tomorrow. At that point, we might want to consider getting the details that the handler throws. It might be ideal to split it as two middleware's.
Adding @dnduffy